octobot_trading/traders/trader.py

```diff
@@ -145,48 +154,122 @@ async def _create_not_loaded_order(self, new_order: Order, portfolio) -> Order:
+        if order is from a real exchange.
+        :param order: Order to cancel
+        :param is_cancelled_from_exchange: When True, will not try to cancel this order on real exchange
+        :param ignored_order: Order not to cancel if found in linked orders recursive cancels (ex: avoid cancelling
+        a filled order)
+        :return: None
+        """
         if order and not order.is_cancelled() and not order.is_filled():
             async with order.lock:
-                odr = order
-                cancelled_order = await odr.cancel_order()
-                self.logger.info(f"{odr.symbol} {odr.get_name()} at {odr.origin_price}"
-                                 f" (ID : {odr.order_id}) cancelled on {self.exchange_manager.exchange_name}")
-
-                # skip_upsert when cancelled_order is None (nothing to update because simulated order)
-                skip_upsert = cancelled_order is None
-                cancelled_order = odr if skip_upsert else cancelled_order
-                await self.exchange_manager.exchange_personal_data.handle_order_update(order.symbol,
-                                                                                       order.order_id,
-                                                                                       cancelled_order,
-                                                                                       skip_upsert=skip_upsert)
+                # always cancel this order first to avoid infinite loop followed by deadlock
+                order.cancel_order()
+                for linked_order in order.linked_orders:
+                    if linked_order is not ignored_order:
+                        await self.cancel_order(linked_order, ignored_order=ignored_order)
+                raw_cancelled_order = await self._handle_order_cancellation(order, is_cancelled_from_exchange)
+                self.logger.info(f"{order.symbol} {order.get_name()} at {order.origin_price}"
+                                 f" (ID : {order.order_id}) cancelled on {self.exchange_manager.exchange_name}")
+
+                # nothing to update in case of a simulated order
+                if raw_cancelled_order is not None:
+                    await self.exchange_manager.exchange_personal_data.handle_order_update_from_raw(order.symbol,
+                                                                                                    order.order_id,
+                                                                                                    raw_cancelled_order)
+
+    async def _handle_order_cancellation(self, order: Order, is_cancelled_from_exchange: bool) -> dict:
+        raw_cancelled_order = None
+        # if real order: cancel on exchange
+        if not self.simulate and not order.is_self_managed() and not is_cancelled_from_exchange:
+            raw_cancelled_order = await self.exchange_manager.exchange.cancel_order(order.order_id, order.symbol)
```

Comment:
Should we check something on the exchange return ?

Reply:
It's the same as when we create an order on exchange, we need to decide where the checks are processed (on whether the command worked or not and ensure the complete order is returned). I think it makes more sense to handle it in rest exchange and maybe raise an exception if there is an issue there ? We need to discuss it I think

Reply:
Yes we will need something like this in the rest exchange. You're right we should add this check with the one concerning the order creation in an other PR.